### PR TITLE
hawkbit: add support for using Range HTTP header

### PR DESCRIPTION
**Pull request by burumaj**
Downloading of firmware updates can now be performed using the Range
HTTP header. This is because some modems don't support downloading files
larger than some set amount when using TLS. For example the nrf9160
modem can only handle transfers of 2 kB when using TLS. Moreover this
also also provides a starting point for resumeable downloads for
hawkbit.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

### CODE REVIEW

**Review by ycsin — CHANGES_REQUESTED** [review]

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 1]
```diff
@@ -1069,6 +1183,31 @@ static bool send_request(enum http_method method,
...
+					      HAWKBIT_RECV_TIMEOUT, "HAWKBIT_DOWNLOAD_PART");
+			if (ret < 0) {
+				LOG_ERR("Unable to send HTTP request (HAWKBIT_DOWNLOAD_PART): %d",
+					ret);
+				return false;
+			}
+			if (hb_context.code_status == HAWKBIT_NETWORKING_ERROR) {
+				return false;
+			}
+			hb_context.dl.parts_downloaded++;
```
Comment (Navin-Sankar) [inline subsys/mgmt/hawkbit/hawkbit.c:1204]: parts_downloaded variable is incremented. but it is not consumed anywhere in the code

File: subsys/mgmt/hawkbit/Kconfig [hunk 2]
```diff
@@ -81,6 +81,31 @@ config HAWKBIT_DDI_SECURITY_TOKEN
...
+config HAWKBIT_DOWNLOAD_PART
+	bool "Download using multiple requests"
+	help
+	  Download firmware using multiple requests. This uses the HTTP Range
+	  header to split up downloads into chunks.
+
+endchoice
+
+config HAWKBIT_DOWNLOAD_PART_SIZE
+	int "Size of downloaded part"
```
[the remark this answers is not in the document]
Comment (Navin-Sankar) [reply subsys/mgmt/hawkbit/Kconfig:102]: default 1 KB is defined

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 3]
```diff
@@ -881,14 +933,76 @@ static void response_cb(struct http_response *rsp,
 			hb_context.final_data_received = true;
 		}
 
+		break;
+	case HAWKBIT_DOWNLOAD_PART:
+		if (rsp->http_status_code != 206) {
```
Comment (Navin-Sankar) [inline subsys/mgmt/hawkbit/hawkbit.c:938]: Add ifdef here for conditional compilation

File: subsys/mgmt/hawkbit/Kconfig (hunk 2 above)
[the remark this answers is not in the document]
Comment (Navin-Sankar) [reply subsys/mgmt/hawkbit/Kconfig:102]: by default 1KB size is defined

[the remark this answers is not in the document]
Comment (ycsin) [reply subsys/mgmt/hawkbit/Kconfig:102]: something like 
```
	int "Size of downloaded part, unit in kB"
```
would be nice

**Comment by ycsin** [comment]
@burumaj do you have any idea how this PR facilitates the resumable download?

For the typical download, I planned to simply record the progress at the time of error and use that to resume the download, the error can happen at any time. The next time when hawkbit want to download the file again, it will check if there was an unfinished download and it will pick up from there. The range would simply be from the last stored progress to the end of the firmware, which I think should be rather straightforward and simple.

**Comment by burumaj** [comment]
@ycsin This just eases that effort, enabling resumable downloads now only requires keeping track of how much of the firmware has been donwloaded. When a download does fail, resuming the download now requires the following steps I think:
1) From the start of slot1, read all downloaded bytes in clean `md_context`
2) set `hb_context.dl.firmware_size` and `hb_context.dl.downloaded_size` to the correct values
3) from `send_request` set `req.header_cb` and pass it `user_data` like `"HAWKBIT_DOWNLOAD_RESUME"`. This would require slightly extending the `header_cb function`
4) Download as usual

File: subsys/mgmt/hawkbit/hawkbit.c (hunk 1 above)
Comment (burumaj) [reply subsys/mgmt/hawkbit/hawkbit.c]: It has been removed

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 4]
```diff
@@ -71,6 +75,7 @@ struct hawkbit_download {
 	int download_progress;
 	size_t downloaded_size;
 	size_t http_content_size;
+	size_t firmware_size;
```
Comment (rlubos) [inline subsys/mgmt/hawkbit/hawkbit.c:78]: Who initializes this newly introduced member? I see it being used in the code, but I don't see it being set to some value at any place. This is specifically concerning given that the value is used as divisor at line 990.

Comment (burumaj) [reply subsys/mgmt/hawkbit/hawkbit.c:78]: It's added now. It's set on line 520 now.

**Review by rlubos — APPROVED** [review]

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 5]
```diff
@@ -1069,6 +1187,30 @@ static bool send_request(enum http_method method,
 		}
 
 		break;
+
+	case HAWKBIT_DOWNLOAD_PART:
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:1191]: The difference of `HAWKBIT_DOWNLOAD_PART` from the usual `HAWKBIT_DOWNLOAD` is basically the addition of `Range` header and a 'tweaked' version of the response's `HAWKBIT_DOWNLOAD` handling to set the `final_data_received` flag when it has downloaded all of the firmware.

Do you think if it is possible to streamline this a little such that there's only one type of download, with the `Range` header there as default, Kconfig is used to limit the download size, if defined, and do a similar loop that you currently have. Then the `response_cb` will handle the response differently, especially the `HTTP_DATA_FINAL`, depending on the Kconfig.

Then we will continue to have a single response_cb to handle the download datasteam, lesser code duplication, and easier to implement resumable download.

File: samples/subsys/mgmt/hawkbit/sample.yaml [hunk 6]
```diff
@@ -10,3 +10,12 @@ tests:
     depends_on: netif
     build_only: true
```
Comment (ww34ww) [inline samples/subsys/mgmt/hawkbit/sample.yaml:11]: Ehh

**Review by ww34ww — APPROVED** [review]

File: subsys/mgmt/hawkbit/hawkbit.c (hunk 5 above)
Comment (ycsin) [reply subsys/mgmt/hawkbit/hawkbit.c:1191]: @Navin-Sankar any thoughts?

**Pull request closed without merging**